app/models/release.rb

```diff
@@ -139,6 +151,16 @@ class Release < ApplicationRecord
       if: :tag?,
     }
 
+  # Assert that package matches the release's product.
+  validate on: %i[create update] do
+    next unless
+      release_package_id_changed?
+
+    unless package.nil? || package.product_id == product_id
+      errors.add :package, :not_allowed, message: 'package product must match release product'
```

Comment:
I don't like the denormalization here, but there's really no way around it since package could be nil (meaning we can't have the release associated with a product through the package).

Reply:
It's essentially the same thing as we have for `account_id` and `environment_id`, so meh.